src/Groups/matrices/transform_form.jl

```diff
@@ -0,0 +1,412 @@
+# TODO: in this file are used many methods TEMPORARILY defined in files matrix_manipulation.jl and stuff_field_gen.jl
+# once methods in those files will be deleted / replaced / modified, this file need to be modified too
+
+
+
+import AbstractAlgebra: Field
+
+export
+    iscongruent
+
+###############################################################################################################
+
+# Algorithm
+
+###############################################################################################################
+
+# The functions find_radical, block_anisotropic_elim and block_herm_elim are based on the article
+# James B. Wilson, Optimal Algorithms of Gram-Schmidt type, Linear Algebra and its Applications 438 (2013) 4573-4583
+
+# returns x,y such that ax^2+by^2 = c
+# at the moment, it simply search for x such that (c-ax^2)/b is a square
+# TODO: is there a faster way to find x and y?
+# TODO: it would be better if this is deterministic. This depends on gen(F) and issquare(F).
+function _solve_eqn(a::T, b::T, c::T) where T <: FieldElem
+   F = parent(a)  
+   ch = Int(characteristic(F))
+   dg = degree(F)
+   w = gen(F)
+   for i in collect(Iterators.product([0:ch-1 for j in 1:dg]...))
+      x = sum([w^(j-1)*i[j] for j in 1:dg])
+      s = (c - a*x^2)*b^-1
+#      vero, y = issquare(s)
+      if issquare(s) return x,square_root(s) end
+   end
+   return nothing, nothing
+end
+
+
+
+# if Symmetric, returns C,A,d where A*B*transpose(frobenius(A,e)) = C, C = block_matrix(2,2,[C,0,0,0]) and d = rank(C)
+# else returns C,A,d where B*A = C, C = [C 0] and d = rank(C)
+# Assumption: if Symmetric==true, then nr=nc always
+function find_radical(B::MatElem{T}, F::Field, nr::Int, nc::Int; e=0, Symmetric=false) where T <: FieldElem
+
+   V1 = VectorSpace(F,nc)
+   V2 = VectorSpace(F,nr)
+   K = Symmetric ? kernel(ModuleHomomorphism(V1,V2,B))[1] : kernel(ModuleHomomorphism(V1,V2,transpose(B)))[1]
+   U,emb = complement(V1,K)
+   d = dim(U)
+   A = matrix(vcat(typeof(emb(gen(U,1)))[emb(v) for v in gens(U)], typeof(K.map(gen(K,1)))[K.map(v) for v in gens(K)] ))
```

Comment:
That line is tough to read. Perhaps you can split it into multiple, assigning some intermediate values to variables with meaningful names. Like `T1 = typeof(emb(gen(U,1)))` (but ideally give `T1 ` a more meaningful name... I'd try to make a suggestion but I am not sure I understand enough... Hmm, but let's see, `emb` is a map and you are computing the type of an element in its codomain, right? So perhaps one should not do this by evaluating `emb`, but rather do something like `T1  = elem_type(codomain(emb))` (assuming this works as I think it does... best to test it?).

Then there is `T2 = typeof(K.map(gen(K,1)))` which I assume is similar... but what is `K.map`? Is it the embedding of the kernel into `V1`? So presumably it is also the second value returned by `kernel` above? Then it might be cleaner to write `K, embK = kernel(...)` above and use `embK` here instead of `K.map` (and rename `emb` to `embU`?)

Meta remark (for @fieker @wbhart @thofma : I have lots of applications where I need to map a vector of elements through a homomorphism (like `phi(gens(G))`). Could we provide this "natively" for convenience? I.e. doing more or less what @GDeFranceschi is doing manually here, mapping a `Vector{srcT}` to a `Vector{dstT}` by mapping each generator separately. Ideally writing this once, in a generic fashion, so that it applies to all homomorphisms (or at least once for each of ring- / group- / module- / ... homemorphisms)

Reply:
In this case, `T1=T2=elem_type(V1)`. I put this correction.